// react-native-valley-rtc-api/RCTValleyVideoView.js
import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {
  requireNativeComponent,
  ViewPropTypes,
  View,
} from 'react-native';

// if ViewPropTypes is not defined fall back to View.propType (to support RN < 0.44)
const viewPropTypes = ViewPropTypes || View.propTypes;

const propTypes = {
  ...viewPropTypes,

  index: PropTypes.number,
  userId: PropTypes.string,
  remove: PropTypes.bool,
  local: PropTypes.bool,
  reload: PropTypes.bool,  // reload the view after reload is true

  /*
    // todo: how to use this method to resolve this trouble
    center: PropTypes.shape({
    latitude: PropTypes.number.isRequired,
    longitude: PropTypes.number.isRequired,
  }).isRequired,
  */
}

let RNValleyRtcApiVideoView = requireNativeComponent('RNValleyRtcApiVideoView', RCTValleyVideoView);
class RCTValleyVideoView extends Component {

  render() {
    return <RNValleyRtcApiVideoView  {...this.props}/>
  }
}

RCTValleyVideoView.propTypes = propTypes;

module.exports = RCTValleyVideoView;
